lib/model.js

```diff
@@ -1588,35 +1588,68 @@ Model.discriminators;
  * @return {Object} the translated 'pure' fields/conditions
  */
 Model.translateAliases = function translateAliases(fields) {
-  if (typeof fields === 'object') {
-    // Fields is an object (query conditions or document fields)
-    for (const key in fields) {
-      let alias;
-      const translated = [];
-      const fieldKeys = key.split('.');
-      let currentSchema = this.schema;
-      for (const field in fieldKeys) {
-        const name = fieldKeys[field];
-        if (currentSchema && currentSchema.aliases[name]) {
-          alias = currentSchema.aliases[name];
-          // Alias found,
-          translated.push(alias);
-        } else {
-          // Alias not found, so treat as un-aliased key
-          translated.push(name);
-        }
+  const translate = (key, value) => {
+    let alias;
+    const translated = [];
+    const fieldKeys = key.split('.');
+    let currentSchema = this.schema;
+    for (const i in fieldKeys) {
+      const name = fieldKeys[i];
+      // console.log(name);
+      if (currentSchema && currentSchema.aliases[name]) {
+        alias = currentSchema.aliases[name];
+        // Alias found,
+        translated.push(alias);
+      } else {
+        // Alias not found, so treat as un-aliased key
+        translated.push(name);
+      }
 
-        // Check if aliased path is a schema
-        if (currentSchema.paths[alias])
-          currentSchema = currentSchema.paths[alias].schema;
-        else
-          currentSchema = null;
+      // Check if aliased path is a schema
+      if (currentSchema.paths[alias]) {
+        currentSchema = currentSchema.paths[alias].schema;
       }
+      else
+        currentSchema = null;
+    }
 
-      const translatedKey = translated.join('.');
-      fields[translatedKey] = fields[key];
-      if (translatedKey !== key)
+    const translatedKey = translated.join('.');
+    if (fields.set)
+      fields.set(translatedKey, value);
+    else
+      fields[translatedKey] = value;
+
+    if (translatedKey !== key)
+      // We'll be using the translated key instead
+      if (fields.delete)
+        // Delete from map
+        fields.delete(key);
+      else
+        // Delete from object
         delete fields[key]; // We'll be using the translated key instead
+    return fields;
+  }
+
+  if (typeof fields === 'object') {
+    // Fields is an object (query conditions or document fields)
+    if (fields.set) {
+      // Infer that a Map was supplied
+      for (const [key, value] of new Map(fields)) {
```

Comment:
No [array destructuring](http://thecodebarbarian.com/an-overview-of-destructuring-assignments-in-node-js) in Node.js 4 or 5, so you'll need to do `for (const key of fields) {}`

Reply:
Ok changed that.